api/core/app/apps/base_app_queue_manager.py

```diff
@@ -51,12 +57,31 @@ def listen(self) -> Generator:
         listen_timeout = current_app.config.get("APP_MAX_EXECUTION_TIME")
         start_time = time.time()
         last_ping_time = 0
-
+        last_message = None
         while True:
             try:
                 message = self._q.get(timeout=1)
                 if message is None:
+                    try:
```

Comment:
Can you explain the purpose of this code snippet? It seems like it will cause a sharp increase in the memory usage of Redis.

Reply:
I think it's more suitable to subscribe to AppGenerateService.generate() for the streaming return content, rather than putting it in the queue manager.

Reply:
<img width="1407" alt="image" src="https://github.com/langgenius/dify/assets/43173467/92e4afb1-1813-49c8-8a9d-eef6ed097b1b">  

Yes, it is indeed possible, but we considered that in the case of multi-container deployment, we need to use a message middleware to cache the messages generated by GPT, and writing to redis is a very fast operation, so we did this

Reply:
Two Issues:
1. All messages are written to redis here, and the expiration time is very long, which will quickly fill up redis abnormally.
2. `AppQueueManager` only deals with asynchronous communication between the main thread and the sub-thread. It is not appropriate to place this code here. I suggest adding a unified processing method in `core.app.task_pipeline.based_generate_task_pipeline.BasedGenerateTaskPipeline`, and adding this logic in the return of G`enerateTaskPipeline.process()` of the app that needs to be processed.